src/workerd/api/js-rpc-target.c++

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2017-2023 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include <workerd/api/js-rpc-target.h>
+#include <workerd/api/global-scope.h>
+
+namespace workerd::api {
+
+kj::Promise<void> JsRpcTargetImpl::call(CallContext callContext) {
+  auto methodName = kj::heapString(callContext.getParams().getMethodName());
+  // TODO(now): Deserialize the arguments.
+  //    - Current demo doesn't take args, once we pass args on the client side I'll deserialize
+  //      on the server side.
+  auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+  // Try to execute the requested method.
+  try {
+    co_await ctx.run(
+        [this,
+         methodName=kj::mv(methodName),
+         serializedArgs = kj::mv(serializedArgs),
+         entrypointName = entrypointName,
+         &callContext] (Worker::Lock& lock) mutable {
+
+      auto reader = callContext.initResults(capnp::MessageSize { 4, 1 }).initResult();
+
+      auto handler = lock.getExportedHandler(entrypointName, ctx.getActor());
+      KJ_IF_MAYBE(h, handler) {
+        jsg::Lock& js = lock;
+
+        auto handle = h->self.getHandle(lock);
+        auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+        auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+        if (!(fnHandle->IsFunction() && !fnHandle->IsPrivate())) {
+          auto errString = "The RPC receiver does not implement the requested method."_kj;
+          reader.setV8Serialized(
+              jsg::serializeV8Rpc(js,
+              jsg::JsValue(js.v8Error(errString))));
+        } else {
+          auto fn = fnHandle.As<v8::Function>();
+          // TODO(now): Currently, no args are passed. Once we get around to sending them from the
+          // client and deserializing here, we should pass them to the function as well.
+          // TODO(now): How do we differentiate throwing an error from JS vs. returning one?
```

Comment:
> How do we differentiate throwing an error from JS vs. returning one?

That kind of depends. I imagine that you'd likely want to wrap this in a `js.tryCatch()` and explicitly handle any synchronously thrown errors. If by "returning one" you mean the call returning a rejected promise, then I think the best thing would be to actually map this into a `jsg::Function<jsg::Promise<jsg::JsRef<jsg::JsValue>>()>` then attach an appropriate error handler to that. Other than having to pass through the `const jsg::TypeHandler<T>&` reference, that's likely the better approach here anyway as it would simplify some of this logic.

Reply:
This is the only thing left (aside from a simple TODO for the serialization methods); I'm mostly unsure how the client code could determine if the exception it received over the wire was thrown or returned.

Maybe both returning and throwing errors should require them to be caught by the worker, in other words if the client receives an exception, it throws it:

```
// Necessary because any exception received over the wire will been thrown.
try {
    stub.mayThrow();
} catch(err) {
    // Handle `err`
}
```

in the worker, regardless of if the remote has defined:

```
class ... {
    mayThrow() { return new Error("Message"); }
}
```
or

```
class ... {
    mayThrow() { throw new Error("Message"); }
}
```

Reply:
I'm sure that there are things I haven't thought of but it occurs to me that you can throw arbitrary types in javascript. So if you throw a string on the server side, you'd expect it to be thrown on the client side. But if you return a string you'd expect it to be returned. In that context, I think an `Error` should be serialized as normal if returned. This also begs the question of what happens if we return a non-serializable type for js rpc (since I cannot swear to you that `Error` is serializable).

Since it's related, I think I'd treat a not okay outcome as whenever we throw for now.

Reply:
Addressed this here: https://github.com/cloudflare/workerd/pull/1028#issuecomment-1806465745